src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -716,19 +718,34 @@ private void CompletePipe()
                 // Return all segments
                 // if _readHead is null we need to try return _commitHead
                 // because there might be a block allocated for writing
-                BufferSegment segment = _readHead ?? _readTail;
-                while (segment != null)
-                {
-                    BufferSegment returnSegment = segment;
-                    segment = segment.NextSegment;
-
-                    returnSegment.ResetMemory();
-                }
+                BufferSegment current = _readHead ?? _readTail;
 
                 _writingHead = null;
                 _readHead = null;
                 _readTail = null;
+
+                if (current != null)
+                {
+                    int index = _pooledSegmentCount;
+                    BufferSegment[] pool = _bufferSegmentPool;
+                    do
+                    {
+                        BufferSegment next = current.NextSegment;
+                        current.ResetMemory();
+
+                        if ((uint)index < (uint)pool.Length)
```

Comment:
ReturnSegment is inlined here but not in `AdvanceReader`?

Reply:
Fixed